specs/plugins.md

```diff
@@ -0,0 +1,134 @@
+- `name` - required, a string identifier for the plugin class, allows developer to ship multiple plugins in a single assembly.
+- `configSection` - optional, a string for which there must be a matching object property at the root of the configuration file. If a plugin supplies a `configSection` and no corresponding property exists in the configuration file an error will be thrown during start up of the proxy.
+- `urlsToWatch` - optional, an array of strings defining the urls for which this plugin will watch instead of the `urlsToWatch` at the root of the config file, this behavior is defined in the [Multi URL support spec](./multi-url-support.md).
+
+```json
+  "plugins": [
+    {
+        "configSection": "randomErrors",
+        "pluginPath": "RandomErrorPlugin\\msgraph-proxy-handler.dll",
+        "name": "RandomErrorPlugin",
+        "urlsToWatch": [
+            "https://graph.microsoft.com/v1.0/*",
+            "https://graph.microsoft.com/beta/*"
+        ],
+    }
+  ],
+  "randomErrors": {
+    "rate": 50,
+    "allowedErrors": [ 429, 500, 502, 503, 504, 507 ]
+  }
+```
+
+Plugins will executed in the order that they are listed in the plugins configuration array. At least one plugin must be present in the configuration, otherwise the proxy will exit with an error informing the user that their configuration is invalid and contains no plugins.
+
+## Plugins
+
+Plugin classes must implement the `IProxyPlugin` interface and provide a parameterless constructor.
+
+```cs
+public interface IProxyPlugin {
+    string Name { get; }
+    void Register(IPluginEvents pluginEvents, IConfigurationSection? configSection);
+}
+
+public interface IPluginEvents {
+    event EventHandler<InitArgs> Init;
+    event EventHandler<OptionsLoadedArgs> OptionsLoaded;
+    event EventHandler<ProxyRequestArgs> Request;
+    event EventHandler<ProxyResponseArgs> Response;
+}
+```
+
+### Registration
+
+The `Register` method of a `IProxyPlugin` will be called after an instance is created and will provide an `ISet<Regex>` to define the urls the plugin wll watch, either using the default or a plugin specific set based on the supplied configuration and an `IConfigurationSection` if one was loaded.
+
+> It is strongly recommended that plugin implementers provide a default configuration in their code.
+
+Plugin classes may register a handler for any events provided in the `IPluginEvents` interface. Plugins registering a handler for the `Init` event should also register a handler for the `OptionsLoaded` event to consume the supplied options.
+
+All of the event arg types will inherit from a common base type which will supply common logic and utilities. A proposed implementation of the `PluginArgsBase` is provided here:
+
+```cs
+public abstract class PluginArgsBase {
+    protected PluginArgsBase(IConsoleLogger consoleLogger) {
```

Comment:
Shall we extend this parameter to `IProxyContext context` (so `public Plugin(IProxyContext context)`), which has a property named `Logger`? That way, we can have a place to pass more information in the future if need be, without breaking existing plugins. Also, for the logger interface, I suggest we keep the name generic, because the implementation decides where the output goes to.